src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js

```diff
@@ -0,0 +1,138 @@
+import { useCallback, useContext, useEffect, useState } from 'react';
+import AppContext from '../../../context/App/AppContext';
+import createCustomerToken from '../../../api/paypalExpress/createCustomerToken';
+import { _objToArray } from '../../../utils';
+import { __ } from '../../../i18n';
+import setPaymentMethodPaypalExpress from '../../../api/paypalExpress/setPaymentMethod';
+import placeOrder from '../../../api/cart/placeOrder';
+import useShippingMethodCartContext from '../../shippingMethod/hooks/useShippingMethodCartContext';
+import LocalStorage from '../../../utils/localStorage';
+import { config } from '../../../config';
+import usePaymentMethodCartContext from '../../paymentMethod/hooks/usePaymentMethodCartContext';
+import useBillingAddressCartContext from '../../billingAddress/hooks/useBillingAddressCartContext';
+
+export default function usePaypalExpress() {
+  const [, appActions] = useContext(AppContext);
+  const { setErrorMessage, setPageLoader } = appActions;
+  const { selectedMethod } = useShippingMethodCartContext();
+  const { cartBillingAddress } = useBillingAddressCartContext();
+  const query = window.location.href.split('?')[1];
+  const [processPaymentEnable, setProcessPaymentEnable] = useState(false);
+  const { selectedPaymentMethod, methodList } = usePaymentMethodCartContext();
+  const cartId = LocalStorage.getCartId();
+
+  /*
+   Check if is possible to proceed on placing the order.
+   */
+  useEffect(() => {
+    if (
+      query &&
+      selectedMethod?.methodCode &&
+      (selectedPaymentMethod?.code === '' ||
+        selectedPaymentMethod?.code === 'paypal_express')
+    )
+      setProcessPaymentEnable(true);
+  }, [query, setProcessPaymentEnable, selectedMethod, selectedPaymentMethod]);
+
+  useEffect(() => {
+    const processPaypalExpress = async () => {
+      setProcessPaymentEnable(false);
+      /*
+       Get paypal express required data from response query
+       */
+      const { token, PayerID: payerId } = getTokenPayerId(query);
+      const paymentMethod = _objToArray(methodList).find(
+        method => method.code === 'paypal_express'
+      );
+
+      if (
+        !token ||
+        !payerId ||
+        !cartId ||
+        !selectedMethod ||
+        !paymentMethod ||
+        !cartBillingAddress
```

Comment:
Instead of passing the whole `cartBillingAddress`, here we need a `boolean` check. this we can do inside `usePaypalExpressCartContext`. the variable can be `hasCartBillingAddress =  billingAddress.firstName && billingAddress.country`. Passing the whole billingAddress is a risk because that state is prone to change from other actions. but making it boolean will definitely avoid re-rendering of `useCallback` that wraps this function

Reply:
Done